RecoParticleFlow/PFProducer/plugins/SimPFProducer.cc

```diff
@@ -0,0 +1,295 @@
+  }
+
+  // make blocks out of calo particles so we can have cluster references
+  // likewise fill out superclusters
+  std::unique_ptr<reco::SuperClusterCollection> superclusters(new reco::SuperClusterCollection);
+  std::unique_ptr<reco::PFBlockCollection> blocks(new reco::PFBlockCollection);
+  std::unordered_map<size_t,size_t> simCluster2Block;
+  std::unordered_map<size_t,size_t> simCluster2BlockIndex;
+  std::unordered_multimap<size_t,size_t> caloParticle2SimCluster;
+  std::vector<int> caloParticle2SuperCluster;
+  for( unsigned icp = 0; icp < CaloParticles.size(); ++icp ) {
+    blocks->emplace_back();
+    auto block = blocks->back();
+    const auto& simclusters = CaloParticles[icp].simClusters();
+    double pttot = 0.0;
+    double etot  = 0.0;
+    std::vector<size_t> good_simclusters;
+    for( unsigned isc = 0; isc < simclusters.size() ; ++isc ) {
+      auto simc = simclusters[isc];
+      edm::Ref<std::vector<reco::PFCluster> > clusterRef(SimClustersH,simc.key());
+      if( clusterRef->energy() > 0.0 ) {	
+	good_simclusters.push_back(isc);
+	etot += clusterRef->energy();
+	pttot += clusterRef->pt();	
+	std::unique_ptr<reco::PFBlockElementCluster> bec( new reco::PFBlockElementCluster(clusterRef,reco::PFBlockElement::HGCAL) );
+	block.addElement(bec.get());
+	simCluster2Block[simc.key()] = icp;
+	simCluster2BlockIndex[simc.key()] = bec->index();
+	caloParticle2SimCluster.emplace(icp,simc.key());
+      }
+    }
+    auto pdgId = std::abs(CaloParticles[icp].particleId());
+    caloParticle2SuperCluster.push_back(-1);
+    if( (pdgId == 22 || pdgId == 11) && pttot > superClusterThreshold_ ) {
+      caloParticle2SuperCluster[icp] = superclusters->size();
+      math::XYZPoint seedpos; // take seed pos as supercluster point
+      reco::CaloClusterPtr seed;
+      reco::CaloClusterPtrVector clusters;
+      for( auto idx : good_simclusters ) {
+	edm::Ptr<reco::PFCluster> ptr(SimClustersH, simclusters[idx].key());
+	clusters.push_back(ptr);
+	if( seed.isNull() || seed->energy() < ptr->energy() ) {
+	  seed = ptr;
+	  seedpos = ptr->position();
+	}
+      }
+      superclusters->emplace_back(etot,seedpos,seed,clusters);
+    }
+  }
+  auto blocksHandle = evt.put(std::move(blocks));
+  auto superClustersHandle = evt.put(std::move(superclusters),"perfect");
+  
+  // list tracks so we can mark them as used and/or fight over them
+  std::vector<bool> usedTrack(TrackCollection.size(),false), 
+                  //usedGsfTrack(GsfTrackCollection.size(),false), 
+                    usedSimCluster(SimClusters.size(),false);
+
+  std::unique_ptr<reco::PFCandidateCollection> candidates(new reco::PFCandidateCollection);
+  // in good particle flow fashion, start from the tracks and go out
+  for( unsigned itk = 0; itk < TrackCollection.size(); ++itk ) {
+    auto tkRef  = TrackCollection.refAt(itk);
+    std::cout << "Track: " << itk << ' ' << tkRef->pt() << ' ' << tkRef->eta() << ' ' << tkRef->phi() << std::endl;
```

Comment:
Note that "cout" is not thread safe and will produce garbled output if the code is running multithreaded.
Better change to "LogInfo" (not printed by MessageLogger by default) or "LogPrint" (printed by MessageLogger by default) to use thread safe version.
We typically allow the pattern ``` if (debug) cout << ... ```, but it should be understood that it has issues and will not allow debugging of a job running with multiple threads.